### PR TITLE
Implement Compact for Withdrawal

### PR DESCRIPTION
**Pull request by ghost**
Ref https://github.com/paradigmxyz/reth/issues/7050

Expanded macro: https://gist.github.com/0xqd/d1864f7c3d45864d155bb1954195b8b5

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
definitely need some help here @joshieDo 

unclear why this type has flags

**Comment by 0xqd** [comment]
> definitely need some help here @joshieDo
> 
> unclear why this type has flags

Because we have u64 type within Withdrawal, it needs the Flags to track compacted len.

**Review by joshieDo — CHANGES_REQUESTED** [review]
> Because we have u64 type within Withdrawal, it needs the Flags to track compacted len.

Yeah exactly


I'd say not to copy paste the expanded code when dealing the bitflags, and actually use the crate `modular_bitfield` to implement WithdrawalFlags.  If unsure on the bitsize, there is a function called `get_bit_size` for the flag size (u64 is 4)

**Review by mattsse — CHANGES_REQUESTED** [review]
this is a bit embarrassing that I haven't thought about this earlier, but we can solve all of this a lot easier by simply moving the reth-primitive type together with it's derive into the codecs crate and do type conversion when we implement the compact trait manually for the alloy type:

```rust
// in codecs/alloy.rs

// have a private type for withdrawal, the same as in primitives with the derive
#[main_codec]
#[derive(Debug, Clone, PartialEq, Eq, Default, Hash, RlpEncodable, RlpDecodable)]
 struct Withdrawal {
    /// Monotonically increasing identifier issued by consensus layer.
    #[serde(with = "u64_hex")]
    pub index: u64,
    /// Index of validator associated with withdrawal.
    #[serde(with = "u64_hex", rename = "validatorIndex")]
    pub validator_index: u64,
    /// Target address for withdrawn ether.
    pub address: Address,
    /// Value of the withdrawal in gwei.
    #[serde(with = "u64_hex")]
    pub amount: u64,
}

impl Compact for alloy::Withdrawal {
  fn to_compact(self,...) {
    // convert self to self::Withdrawal
    // encode this
 }

}

```

**Comment by 0xqd** [comment]
yeah, it can be one approach using type wrapper and use it to convert internally in Compact. I think the end result would be something like I suggested you last week regarding type wrapper without duplicate the struct from alloy primitives. Btw, should we now just clean up the Flags implementation and move to drop reth primitives Log?

**Comment by mattsse** [comment]
no wrapper, just a a private duplicated type that's used for compact.
this way we can still get rid of reth-primitives::Withdrawal

**Comment by 0xqd** [comment]
@mattsse can you help checking new implementation?

File: crates/storage/codecs/src/alloy/withdrawal.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        buf = new_buf;
+        let obj = Withdrawal { index, validator_index, address, amount };
+        (obj, buf)
+    }
+}
+
+/// Flags to store metadata about compacted Withdrawal.
+pub(crate) struct WithdrawalFlags {
+    bytes: [u8; 2], // Calculated from B4 Specifier
+}
```
Comment (joshieDo) [inline crates/storage/codecs/src/alloy/withdrawal.rs:52]:
```
use modular_bitfield::prelude::*;
                #[doc = "Fieldset that facilitates compacting the parent type. Used bytes: 2 | Unused bits: 1"]
                #[bitfield]
                #[derive(Clone, Copy, Debug, Default)]
                pub struct TestStructFlags {
                    pub f_u64_len: B4,
                    pub f_u256_len: B6,
                    pub f_bool_t_len: B1,
                    pub f_bool_f_len: B1,
                    pub f_option_none_len: B1,
                    pub f_option_some_len: B1,
                    pub f_option_some_u64_len: B1,
                    #[skip]
                    unused: B1,
                }
```

If you implement it as above (taken from one of our tests), all the impl below will be auto generated

**Review by mattsse — CHANGES_REQUESTED** [review]
it's unclear to me why this type needs flags @joshieDo ?

but I don't want to do any of this manually,
we should duplicate the reth-primitive type here and do simple conversions instead this way we can use the derived code.

does that make sense?

basically duplicating the Withdrawal type with derives in the codecs crate and do a conversion in the Compact for alloy::Withdrawal impl

**Comment by 0xqd** [comment]
yes it would be much simpler, let me implement it.

**Review by joshieDo — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
nice, next we can replace the existing type

**Pull request merged**